xls/modules/dbe/lz4_encoder.x

```diff
@@ -0,0 +1,832 @@
+                ..zero_tok
+            })
+        } else if (
+            cur.fsm == FsmSt::START_MATCH_1 &&
+            (!is_match || upd.finalize)
+        ) {
+            // Emit symbol at OP as literal
+            (true, Token{
+                kind: TokenKind::LT,
+                lt_sym: upd.fifo_in[0],
+                ..zero_tok
+            })
+        } else if cur.fsm == FsmSt::CONTINUE_MATCH_1 && is_match_terminated {
+            // If match is long enough, emit a single CP
+            // If not, emit symbols from Cache FIFO as literals
+            if match_is_long {
+                let off = upd.pnew - upd.pold;
+                let t = Token {
+                    kind: TokenKind::CP,
+                    cp_off: off - uN[PTR_WIDTH]:1,
+                    cp_cnt: (match_len - u32:1) as uN[CNT_WIDTH],
+                    ..zero_tok
+                };
+                (true, t) 
+            } else {
+                (true, Token {
+                    kind: TokenKind::LT,
+                    lt_sym: upd.fifo_cache[upd.match_nconts - u32:1],
+                    ..zero_tok
+                })
+            }
+        } else if cur.fsm == FsmSt::EMIT_SHORT_MATCH {
+            (true, Token {
+                kind: TokenKind::LT,
+                lt_sym: upd.fifo_cache[upd.match_nconts - u32:1],
+                ..zero_tok
+            })
+        } else if cur.fsm == FsmSt::EMIT_FINAL_LITERALS {
+            // We dump literals from Input FIFO till nvalid becomes 0
+            (true, Token {
+                kind: TokenKind::LT,
+                lt_sym: upd.fifo_in[u32:0],
+                ..zero_tok
+            })
+        } else if cur.fsm == FsmSt::EMIT_END {
+            (true, Token {
+                kind: TokenKind::MK,
+                mark: Mark::END,
+                ..zero_tok
+            })
+        } else {
+            (false, zero_tok)
+        };
+        let tok_out = send_if(tok, o_encoded, do_emit, encoded_tok);
+
+        // Handle state re-initialization
+        let upd = if cur.fsm == FsmSt::RESET {
+            // Full reset
+            // NOTE: .fsm value will be overridden by the state change logic
+            init_state<SYM_WIDTH, PTR_WIDTH, CNT_WIDTH, HASH_BITS,
+                        MINMATCH, FINAL_LITERALS, FIFO_CACHE_SZ, FIFO_IN_SZ>()
+        } else if cur.fsm == FsmSt::RESTART {
+            // Intra-block partial reset, keeping HB and HT intact
+            State {
+                fifo_in_count: u32:0,
+                fifo_in_nvalid: u32:0,
+                match_nconts: u32:0,
+                finalize: false,
+                recycle: false,
+                ..upd
+            }
+        } else {
+            upd
+        };
+
+        // State change logic
+        let fsm = cur.fsm;
+        let fsm = if rx_error || rx_unexpected_mark {
+            FsmSt::ERROR
+        } else if rx_reset {
+            FsmSt::RESET
+        } else if fsm == FsmSt::RESET {
```

Comment:
Nit: usually we prefer a `match` statement for FSMs in the DSL (so you can demonstrate that you exhaustively handled all cases in a declarative fashion).

Reply:
Fixed.